### ISSUE TITLE
CS8669 in Razor generated code

### ISSUE THREAD

**Issue by drauch** [dotnet/aspnetcore#40866]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

When switching to .NET 6 ReSharper suggested that some of my null checks in Razor files are not necessary because my Model is never null. That's not true though, which is why I changed the model declaration to a nullable one, e.g.:

```
@model string?
```

However, when I compile my solution with warnings as errors and use Razor runtime compilation I get the following errors when executing this Razor page:

> tylng5v1.afl(32,109): error CS8669: The annotation for nullable reference types should only be used in code within a '#nullable' annotations context. Auto-generated code requires an explicit '#nullable' directive in source.

I tried to circumvent this problem with the following idea:

```
@{
  #nullable enable
}
@model string?
```

However, it still doesn't work.

Any ideas on what I am doing wrong? I've found https://github.com/dotnet/aspnetcore/issues/37510 but it says everthing is fixed with 6.0.102 and I'm using an up-to-date VS 2022 with an SDK version of 6.0.201 (64bit).

### Expected Behavior

Nullable models in Razor pages work without warnings.

### .NET Version

6.0.201

### Anything else?

Up-to-date ASP.NET Core 6.x
Up-to-date VS 2022 Professional

**Comment by mkArtakMSFT** [comment]
Thank you for filing this issue. In order for us to investigate this issue, please provide a minimalistic repro project (ideally a GitHub repo) that illustrates the problem.

**Comment by drauch** [comment]
@mkArtakMSFT : Here is a solution that demonstrates the problem:

https://www.dropbox.com/s/1iv79foz9luoa2r/ReproAspNetCore40866ModelNullabilityProblems.zip?dl=0

**Comment by adams85** [comment]
I can confirm that the issue exists in the latest .NET release (6.0.3). I also ran into this when I was upgrading a project from .NET Core 3.1 to .NET 6.

It works fine when runtime compilation is off but I get the same error when I turn it on. In my project `TreatWarningsAsErrors` is enabled too. When I disable that, the compilation error goes away.

So it's clear that the issue is related to `TreatWarningsAsErrors`. After some digging in the sources, I found that [the Razor compiler picks up this setting from the project](https://github.com/dotnet/aspnetcore/blob/v6.0.3/src/Mvc/Mvc.Razor.RuntimeCompilation/src/CSharpCompiler.cs#L198).

I guess this warning is some recent addition to Roslyn, which blocks the runtime Razor compilation when `TreatWarningsAsErrors` is enabled as the Razor engine doesn't emit the explicit `#nullable enable` into the compilation output.

@drauch As a hotfix, you can disable `TreatWarningsAsErrors`. I didn't like this idea though, so I came up with another [workaround](https://github.com/adams85/aspnetskeleton2/blob/master/src/UI.AspNetCore/Infrastructure/Razor/CustomRazorRuntimeCompilationMvcBuilderExtensions.cs). This approach emits a `#pragma warning disable 8669` line in the compilation output and that solves the issue for me. (At first I tried `#nullable enable` but interestingly that didn't work...)

**Comment by drauch** [comment]
As far as I can see I do not have runtime compilation turned on in my sample above. Is it on by default?

**Comment by adams85** [comment]
No, it's not on by default either in the `dotnet new` templates or in your repro project. [Here](https://docs.microsoft.com/en-us/aspnet/core/mvc/views/view-compilation?view=aspnetcore-6.0&tabs=visual-studio) it is how you enable it.

What's even more puzzling is that I tested the repro project you submitted and it works for me without errors, although I'm on SDK v6.0.201 too... I needed to make the following changes to produce the error: https://github.com/adams85/aspnetcore-40866-repro/commit/16e7371aa98abfb323fc403ab60d89154d786bfc

**Comment by drauch** [comment]
That's very weird. I'm also on Win10 for what it's worth. I think we will have to wait for officials...
